### DISCUSSION TITLE
Animation support and platform specific design request

### DISCUSSION THREAD

**Discussion by Ethan1983** [cashapp/redwood#912] [Q&A] [answered]
Using Native UI toolkits and yet writing once with quick iterative development is promising and is a step above both Flutter and Compose multi-platform. However, would appreciate some clarity for the following,

- Is there a plan to support animations? How would it work wrt platform specific native implementation?
- Can Redwood support a platform specific customization for a component or page? Assuming initial Redwood adoption for the entire app and design team wanting a customization just for one platform. Is it possible or is the expectation for developers to change the level at which `RedwoodContent` is being applied?

```
setContent {
      CounterTheme {
        RedwoodContent(factories) {
          Counter()
        }
      }
    }
```

**Comment by JakeWharton** [comment]
> Is there a plan to support animations? How would it work wrt platform specific native implementation?

Anything we did would likely be entirely declarative rather than the imperative-style of Compose UI. That is, we can let you choose an animation to run and run it directly within the native UI toolkit, but we're unlikely to offer a time-based system where you control it from within Compose.

This is exacerbated by the fact that the things you would normally want to animate often are not even exposed by the widgets as defined in your schema.

> Can Redwood support a platform specific customization for a component or page? Assuming initial Redwood adoption for the entire app and design team wanting a customization just for one platform.

Hmm customize how? Can you elaborate a bit?

You are already free to implement the widget interfaces differently on each platform depending on what makes sense for that platform. The goal is not to look and feel 100% the same across platforms, but to look and feel as if each implementation belongs on the specific platform it is written for.

Similarly you can use runtime logic to detect the platform on which you're running and change the behavior that way. We don't offer this out-of-the-box but probably will soon.

**Reply by Ethan1983** [reply]
> Anything we did would likely be entirely declarative rather than the imperative-style of Compose UI. That is, we can let you choose an animation to run and run it directly within the native UI toolkit.

Gotcha, and would this too be implemented in a similar interface pattern for each platform?

> Hmm customize how? Can you elaborate a bit?

Say, design want a specific page layout for a foldable Androids and want to introduce two pane layouts etc. Assuming, this is what you meant by runtime platform logic in presenter code?

```
Column {
    Button("-1", onClick = { count-- })
    Text("Count: $count")
    Button("+1", onClick = { count++ })
  }
```

**Reply by JakeWharton** [reply, marked as answer]
> would this too be implemented in a similar interface pattern for each platform?

I assume it would be something pretty basic like a string with common values ("crossfade", "slide-in", etc.).

> design want a specific page layout for a foldable Androids and want to introduce two pane layouts

It's not implemented yet, but #393 will eventually include things such as the device size to allow you to modify the layout that's shown in the same way you would do so with Compose UI and querying the device size.
